src/core/service.c

```diff
@@ -2073,13 +2080,28 @@ static void service_enter_dead(Service *s, ServiceResult f, bool allow_restart)
 
                 log_unit_debug(UNIT(s), "Next restart interval calculated as: %s", FORMAT_TIMESPAN(restart_usec_next, 0));
 
+                /* If the relevant option is set, and the unit doesn't already have logging level set to
+                 * debug, enable it now. Make sure to overwrite the state in /run/systemd/units/ too, to
+                 * ensure journald doesn't prune the messages. The previous state is saved and restored
+                 * once the auto-restart flow ends. */
+                if (s->restart_mode == SERVICE_RESTART_MODE_DEBUG && !UNIT(s)->debug_invocation) {
+                        UNIT(s)->debug_invocation = true;
+                        unit_overwrite_log_level_max(UNIT(s), LOG_PRI(LOG_DEBUG));
+                }
+
                 service_set_state(s, SERVICE_AUTO_RESTART);
-        } else
+        } else {
                 /* If we shan't restart, the restart counter would be flushed out. But rather than doing that
                  * immediately here, this is delegated to service_start(), i.e. next start, so that the user
                  * can still introspect the counter. */
                 service_set_state(s, end_state);
 
+                if (s->restart_mode == SERVICE_RESTART_MODE_DEBUG && UNIT(s)->debug_invocation) {
+                        UNIT(s)->debug_invocation = false;
+                        unit_overwrite_log_level_max(UNIT(s), s->exec_context.log_level_max);
```

Comment:
it would be good to add a log message here, i.e. "Service failed, subsequent restarts will be executed as debug invocations". or something like that.

Reply:
at info LOG_NOTICE level i guess